crates/ibc/src/core/ics03_connection/handler/conn_open_confirm.rs

```diff
@@ -6,10 +6,101 @@ use crate::core::ics03_connection::error::Error;
 use crate::core::ics03_connection::events::OpenConfirm;
 use crate::core::ics03_connection::handler::{ConnectionIdState, ConnectionResult};
 use crate::core::ics03_connection::msgs::conn_open_confirm::MsgConnectionOpenConfirm;
+use crate::core::ics24_host::path::ConnectionsPath;
+use crate::core::{ExecutionContext, ValidationContext};
 use crate::events::IbcEvent;
 use crate::handler::{HandlerOutput, HandlerResult};
 use crate::prelude::*;
 
+pub(crate) fn validate<Ctx>(ctx_b: &Ctx, msg: MsgConnectionOpenConfirm) -> Result<(), Error>
```

Comment:
I have been thinking about how we can avoid the code duplication we currently have and it seems roughly speaking each of these handlers are composed of the following steps - 

validate:
	* get relevant state from host
	* check current state is as expected (based on message)
	* construct expected state
	* verify proofs

execute:
	* get relevant state from host
	* emit event/log
	* construct new state
	* write state

process:
	* get relevant state from host
	* check current state is as expected (based on message)
	* construct expected state
	* verify proofs
	* emit event/log
	* construct new state
	* (write state later on in `store_connection_result`)

So there is some degree of overlap between them. Here's a potential way to avoid the duplication by defining a separate 'handler' trait and providing blanket impls for types that implement the `ValidationContext` and `ConnectionReader` -> 

```rust
trait ConnOpenConfirmHandler {
    fn verify(&self, msg: &MsgConnectionOpenConfirm) -> Result<(), Error> {
        let conn_end_on_b = self.connection_end(&msg.conn_id_on_b)?;
        if !conn_end_on_b.state_matches(&State::TryOpen) {
            return Err(Error::connection_mismatch(msg.conn_id_on_b.clone()));
        }

        let client_id_on_a = conn_end_on_b.counterparty().client_id();
        let client_id_on_b = conn_end_on_b.client_id();
        let conn_id_on_a = conn_end_on_b
            .counterparty()
            .connection_id()
            .ok_or_else(Error::invalid_counterparty)?;

        // Verify proofs
        {
            let client_state_of_a_on_b = self
                .client_state(client_id_on_b)
                .map_err(|_| Error::other("failed to fetch client state".to_string()))?;
            let consensus_state_of_a_on_b = self
                .consensus_state(client_id_on_b, msg.proof_height_on_a)
                .map_err(|_| Error::other("failed to fetch client consensus state".to_string()))?;

            let prefix_on_a = conn_end_on_b.counterparty().prefix();
            let prefix_on_b = self.commitment_prefix();

            let expected_conn_end_on_a = ConnectionEnd::new(
                State::Open,
                client_id_on_a.clone(),
                Counterparty::new(
                    client_id_on_b.clone(),
                    Some(msg.conn_id_on_b.clone()),
                    prefix_on_b,
                ),
                conn_end_on_b.versions().to_vec(),
                conn_end_on_b.delay_period(),
            );

            client_state_of_a_on_b
                .verify_connection_state(
                    msg.proof_height_on_a,
                    prefix_on_a,
                    &msg.proof_conn_end_on_a,
                    consensus_state_of_a_on_b.root(),
                    conn_id_on_a,
                    &expected_conn_end_on_a,
                )
                .map_err(Error::verify_connection_state)?;
        }

        Ok(())
    }

    fn client_state(&self, client_id: &ClientId) -> Result<Box<dyn ClientState>, Error>;

    fn connection_end(&self, conn_id: &ConnectionId) -> Result<ConnectionEnd, Error>;

    fn consensus_state(
        &self,
        client_id: &ClientId,
        height: Height,
    ) -> Result<Box<dyn ConsensusState>, Error>;

    fn commitment_prefix(&self) -> CommitmentPrefix;
}

struct ValidationCtx<'a, T>(&'a T);

impl<T> ConnOpenConfirmHandler for ValidationCtx<'_, T> where T: ValidationContext {
    fn client_state(&self, client_id: &ClientId) -> Result<Box<dyn ClientState>, Error> {
        self.0.client_state(client_id)
            .map_err(|_| Error::other("failed to fetch client state".to_string()))
    }

    fn connection_end(&self, conn_id: &ConnectionId) -> Result<ConnectionEnd, Error> {
        self.0.connection_end(conn_id)
    }

    fn consensus_state(&self, client_id: &ClientId, height: Height) -> Result<Box<dyn ConsensusState>, Error> {
        self.0.consensus_state(client_id, height)
            .map_err(|_| Error::other("failed to fetch consensus state".to_string()))
    }

    fn commitment_prefix(&self) -> CommitmentPrefix {
        self.0.commitment_prefix()
    }
}

struct Ics26Ctx<'a, T>(&'a T);

impl<T> ConnOpenConfirmHandler for Ics26Ctx<'_, T> where T: ConnectionReader {
    fn client_state(&self, client_id: &ClientId) -> Result<Box<dyn ClientState>, Error> {
        self.0.client_state(client_id)
    }

    fn connection_end(&self, conn_id: &ConnectionId) -> Result<ConnectionEnd, Error> {
        self.0.connection_end(conn_id)
    }

    fn consensus_state(&self, client_id: &ClientId, height: Height) -> Result<Box<dyn ConsensusState>, Error> {
        self.0.client_consensus_state(client_id, height)
    }

    fn commitment_prefix(&self) -> CommitmentPrefix {
        self.0.commitment_prefix()
    }
}
```

Reply:
As discussed, I opened #271 where we can further discuss this. Let's focus on correctness in this PR